components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/JsonSchemaValidator.java

```diff
@@ -0,0 +1,188 @@
+import org.wso2.carbon.apimgt.gateway.threatprotection.utils.ThreatProtectorConstants;
+import org.wso2.carbon.apimgt.gateway.utils.GatewayUtils;
+import org.wso2.carbon.apimgt.impl.APIConstants;
+
+import javax.xml.stream.XMLStreamException;
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Map;
+
+/**
+ * This mediator would protect the backend resources from the JSON threat vulnerabilities by validating the
+ * JSON schema.
+ */
+public class JsonSchemaValidator extends AbstractMediator {
+
+    /**
+     * This mediate method validates the message body.
+     *
+     * @param messageContext This message context contains the request message properties of the relevant
+     *                       API which was enabled the JSON_Validator message mediation in flow.
+     * @return a boolean true if the message content is passed the json schema criteria.
+     */
+    public boolean mediate(MessageContext messageContext) {
+        if (log.isDebugEnabled()) {
+            log.debug("JSON schema validation mediator is activated...");
+        }
+        Map<String, InputStream> inputStreams = null;
+        org.apache.axis2.context.MessageContext axis2MC;
+        Boolean validRequest = true;
+        String contentType;
+        String apiContext;
+        String requestMethod;
+        axis2MC = ((Axis2MessageContext) messageContext).getAxis2MessageContext();
+        contentType = axis2MC.getProperty(ThreatProtectorConstants.CONTENT_TYPE).toString();
+        apiContext = messageContext.getProperty(ThreatProtectorConstants.API_CONTEXT).toString();
+        requestMethod = axis2MC.getProperty(ThreatProtectorConstants.HTTP_REQUEST_METHOD).toString();
+
+        if (!APIConstants.SupportedHTTPVerbs.GET.name().equalsIgnoreCase(requestMethod) &&
+                (ThreatProtectorConstants.APPLICATION_JSON.equals(contentType) ||
+                        ThreatProtectorConstants.TEXT_JSON.equals(contentType))) {
+            JSONConfig jsonConfig = configureSchemaProperties(messageContext);
+            ConfigurationHolder.addJsonConfig(jsonConfig);
+            APIMThreatAnalyzer apimThreatAnalyzer = AnalyzerHolder.getAnalyzer(contentType);
+            try {
+                inputStreams = GatewayUtils.cloneRequestMessage(messageContext);
+                if (inputStreams != null) {
+                    InputStream inputStreamJson = inputStreams.get(ThreatProtectorConstants.JSON);
+                    BufferedInputStream bufferedInputStream = new BufferedInputStream(inputStreamJson);
+                    apimThreatAnalyzer.analyze(bufferedInputStream, apiContext);
+                }
+            } catch (APIMThreatAnalyzerException e) {
+                validRequest = false;
+                String message = "Request is failed due to JSON schema validation failure: ";
+                GatewayUtils.handleThreat(messageContext, ThreatProtectorConstants.HTTP_SC_CODE,
```

Comment:
Is this exception is getting logged somewhere? If we are only passing the relevant error message to front-end, how the server knows about this? and here, since we are not passing e. We are going to miss the exception stack trace, which will be a problem later in debugging ?

Reply:
yes.The exception is logged in custom fault sequence.